Add file locking mechanism to manage application instances

DiTsi opened the pull request:
* Introduced FileLock class to handle locking logic
* Updated lifespan management in main.py to utilize file locking
* Implemented heartbeat and unlock waiting functionality for concurrent instance handling

DiTsi committed:
Add file locking mechanism to manage application instances

* Introduced FileLock class to handle locking logic
* Updated lifespan management in main.py to utilize file locking
* Implemented heartbeat and unlock waiting functionality for concurrent instance handling

DiTsi committed:
update test_main

DiTsi committed:
added timeout for tests

DiTsi committed:
Worked HA

DiTsi committed:
Update docs

DiTsi committed:
Merge branch 'develop' into 341-add-ha-support

DiTsi committed:
Fix some issues

tansdf commented on app/file_lock.py:
```diff
@@ -0,0 +1,68 @@
+            except (ValueError, FileNotFoundError) as e:
+                logger.debug(f"Error reading lock file: {e}")
+                return True
+        return False
```
There's a race condition between checking if the lock exists and acquiring it (Time-of-Check-Time-of-Use). Two instance that run simultaneously could both see a free lock. It's better to use atomic file operations with exclusive flags

tansdf commented on app/file_lock.py:
```diff
@@ -0,0 +1,68 @@
+    def _update(self):
+        hostname = socket.gethostname()
+        with open(self.lock_path, "w") as f:
+            f.write(f"{hostname},{os.getpid()},{time.time()}")
```
MIssing file operation error handling. At least we could catch `OSError, IOError`

tansdf commented on app/file_lock.py:
```diff
@@ -0,0 +1,68 @@
+                    return False
+            except (ValueError, FileNotFoundError) as e:
+                logger.debug(f"Error reading lock file: {e}")
+                return True
```
Returning True on error is good and conservative, but can cause issues when the file is corrupted. It's better to return False on `FileNotFoundError` and try to recreate file on `ValueError` I guess

DiTsi replied:
Recheck please new code

tansdf commented on main.py:
```diff
@@ -132,11 +173,79 @@ async def lifespan(fastapi_app: FastAPI):
+    )
+
+
+class StandbyMiddleware(BaseHTTPMiddleware):
```
First of all it should be placed in a separate file

tansdf commented on main.py:
```diff
@@ -132,11 +173,79 @@ async def lifespan(fastapi_app: FastAPI):
+                return await call_next(request)
+            
+            # Block all other requests
+            return service_unavailable_response("Service Unavailable - Standby mode. Only /ready endpoint is available.")
```
Seems the middleware will block all the requests in standby mode, even those in the middle of execution, am I right? Probably it's a good idea to allow running requests to finish but not allow new ones

tansdf commented on app/file_lock.py:
```diff
@@ -0,0 +1,68 @@
+from app.logging import logger
+
+
+class FileLock:
```
Please add unit-tests, at least for logic blocks

tansdf commented on app/file_lock.py:
```diff
@@ -0,0 +1,68 @@
+        self._active = False
+        self._heartbeat_task = None
+
+    def lock(self):
```
It's better to rename to `acquire_lock`

tansdf commented on app/file_lock.py:
```diff
@@ -0,0 +1,68 @@
+        loop = asyncio.get_running_loop()
+        self._heartbeat_task = loop.create_task(self._heartbeat())
+
+    def unlock(self):
```
It's better to rename to `release_lock`

tansdf reviewed (changes requested):
We have failig tests:
```
FAILED tests/test_main.py::TestMainApplication::test_lifespan_startup - TypeError: cannot unpack non-iterable Mock object
FAILED tests/test_main.py::TestMainApplication::test_lifespan_startup_with_locked_file - TypeError: cannot unpack non-iterable Mock object
```
Also we have significant coverage drop for the main.py `44%` - before it was `69%`.

DiTsi committed:
Added new tests and fixed exists

DiTsi committed:
Refactor file locking and add standby middleware

- Renamed `lock` and `unlock` methods in `FileLock` class to `acquire_lock` and `release_lock` for clarity.
- Introduced `StandbyMiddleware` to manage request handling during standby mode, allowing only the `/ready` endpoint.
- Moved middleware-related functions to a new `middleware.py` file for better organization.
- Updated the main application to include the new middleware and ensure proper server initialization and shutdown logging.

DiTsi committed:
Update .dockerignore and .gitignore, and refactor file locking logic

- Added .coverage and htmlcov to .dockerignore and .gitignore for better exclusion of test artifacts.
- Refactored the is_locked and get_lock_info methods in the FileLock class to simplify error handling and improve readability.

DiTsi committed:
Update GitHub workflows to allow configurable runner type

- Modified the feature workflow to accept a runner input, defaulting to 'self-hosted'.
- Updated the develop workflow to specify 'ubuntu-24.04' as the runner for the tests job.

DiTsi committed:
Check CI/CD on self-hosted runner

DiTsi committed:
Check CI/CD on self-hosted runner

DiTsi committed:
Check CI/CD on self-hosted runner

DiTsi committed:
Check CI/CD on self-hosted runner

DiTsi committed:
Trying to run tests with Docker

DiTsi committed:
Trying to run tests with Docker

DiTsi committed:
Fix tests

DiTsi committed:
Switch to podman

DiTsi committed:
Switch to podman

DiTsi committed:
Lock with creating .lock.d directory

DiTsi committed:
Update Chart resources, add deployment strategy, update HA docs

DiTsi committed:
update tests

DiTsi committed:
update tests

DiTsi committed:
Merge branch 'develop' into 341-add-ha-support

DiTsi committed:
remove extra logs

DiTsi committed:
update main.py and its tests

DiTsi committed:
update telegram tests
